### PR TITLE
Handle zap_add() failures in "casesensitivity=mixed" mode.

### PR DESCRIPTION
**Pull request by sanjeevbagewadi**
With "casesensitivity=mixed", zap_add() could fail when the number of
files/directories with the same name (varying in case) exceed the
capacity of the leaf node of a Fatzap. This results in a ASSERT()
failure as zfs_link_create() does not expect zap_add() to fail. The fix
is to handle these failures and rollback the transactions.

Signed-off-by: Sanjeev Bagewadi <sanjeev.bagewadi@gmail.com>

### Description

For a dataset with "casesensitivity=mixed", when a large number of files/directories
with same name (varying only in case e.g: ABCD, ABCd, ABcD and so on) are created
zap_add() could fail. With mixed mode zap_add() normalises the names before the hash
is computed. And all the names would generate the same hash and land in the same leaf.
When the number of entries exceed the capacity of the leaf-block, zap_add() tries to split
the leaf-block which fails as well and zap_add() fails. This trips an ASSERT in zfs_link_create()
as it does not expect zap_add() to fail.

The fix does the following :
- fzap_add_cd() : Handle the case when zap_expand_leaf() fails with ENOSPC and bailout
   without calling zap_put_leaf_maybe_grow_ptrtbl(). 
- zap_add_impl() : When adding to a micro-zap check if the total number of entries
  with colliding/same hash value can fit into fatzap-leaf-block. This is important because, if/when
  the microzap needs to be upgraded to fatzap, all the entries with the same hash would need to
  fit into the same leaf-block (16K). If the number of such entries donot fit fail the zap_add().
  
   The routine mze_canfit_fzap_leaf() today assumes the MZAP_NAME_LEN for every entry.
   This is erring on the safer side but, ends up accommodating lesser number (127) of entries
    with same hash value in microzap. We could find out the size of name of every mze and that
    would be accurate. But, it is expensive to compute the length every time. Alternatively, we
    could compute the length of each entry and cache it. I felt that the amount of code needed
    for this is not worth the gain. I am open to changing it if necessary.

- zfs_link_create() : Move the call to zap_add() to the beginning and in case of a failure
  return. This ensures that we can bailout easily before making any other modifications to
  the parent-zap or the child-dnode. Keeps the code simpler.
- ZPL interfaces (zfs_create(), zfs_mkdir(), zfs_symlink()) : Handle the failure of zfs_link_create()
  and rollback the operation.

With these changes a call to create a file could fail with ENOSPC. Not the best error value.
But, this is the closest I found. Any alternate suggestions are welcome.

### Motivation and Context

With "casesensitivity=mixed" it is easy to panic the node with a simple test case
as described in https://github.com/zfsonlinux/zfs/issues/7011

### How Has This Been Tested?

The following tests were run : 
- zfs-testsuite
- ztest
- Unit-test described in the #7011 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: module/zfs/zap_micro.c [hunk 1]
```diff
@@ -363,6 +363,43 @@ mze_find_unused_cd(zap_t *zap, uint64_t hash)
...
+
+	mze_tofind.mze_hash = hash;
+	mze_tofind.mze_cd = 0;
+
+	for (mze = avl_find(avl, &mze_tofind, &idx);
+	    mze && mze->mze_hash == hash; mze = AVL_NEXT(avl, mze)) {
+		/*
+		 * Each mzap entry requires at max : 4 chunks
+		 * 3 chunks for names + 1 chunk for value.
+		 */
```
Comment (ahrens) [inline module/zfs/zap_micro.c]: This comment seems out of place; seems like it is relevant to MZAP_ENT_CHUNKS.

File: module/zfs/zap_micro.c [hunk 2]
```diff
@@ -1202,6 +1239,10 @@ zap_add_impl(zap_t *zap, const char *key,
 		if (mze != NULL) {
 			err = SET_ERROR(EEXIST);
 		} else {
+			if (!mze_canfit_fzap_leaf(zn, zn->zn_hash)) {
+				zap_unlockdir(zap, tag);
+				return (ENOSPC);
```
Comment (ahrens) [inline module/zfs/zap_micro.c]: Did you consider upgrading to a fat zap in this case?  That would mitigate the slightly-smaller-than-necessary limit on collisions in the microzap due to your overestimation of the name lengths.

File: module/zfs/zfs_dir.c [hunk 3]
```diff
@@ -776,6 +776,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
...
+	if (error != 0) {
+		if (!(flag & ZRENAMING) && !(flag &ZNEW)) {
+			drop_nlink(ZTOI(zp));
+		}
+		/*
+		 * The caller of this routine is responsible for failing the
+		 * transaction which will rollback the SA updates done above.
+		 */
+		mutex_exit(&zp->z_lock);
+		return (error);
```
Comment (ahrens) [inline module/zfs/zfs_dir.c]: need to update the comment on line 745.

File: module/zfs/zfs_vnops.c [hunk 4]
```diff
@@ -1445,10 +1446,25 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
...
 		zfs_mknode(dzp, vap, tx, cr, 0, &zp, &acl_ids);
 
+		error = zfs_link_create(dl, zp, tx, ZNEW);
+		if (error != 0) {
+			/*
+			 * Delete the node created above.
+			 */
+			zp->z_unlinked = 1;
+			drop_nlink(ZTOI(zp));
+			zfs_unlinked_add(zp, tx);
```
Comment (ahrens) [inline module/zfs/zfs_vnops.c]: adding this to the unlinked set seems like a roundabout way of removing it.  Can we just call zfs_znode_delete()?  I know zfs_inactive->zfs_zinactive->zfs_rmnode has some additional logic, but I don't think that it applies in this case.  E.g. this znode can not have an xattrdir at this point (I think).

The same applies to symlink and mkdir.

File: module/zfs/zap_micro.c (hunk 2 above)
Comment (sanjeevbagewadi) [reply module/zfs/zap_micro.c]: I think that makes sense. Will fix it.

File: module/zfs/zap_micro.c (hunk 1 above)
Comment (sanjeevbagewadi) [reply module/zfs/zap_micro.c]: You are right. Will fix it.

File: module/zfs/zfs_dir.c (hunk 3 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_dir.c]: Will do.

File: module/zfs/zfs_vnops.c (hunk 4 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Will do that.

File: module/zfs/zfs_vnops.c [hunk 5]
```diff
@@ -1429,6 +1429,7 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
 			dmu_tx_hold_write(tx, DMU_NEW_OBJECT,
 			    0, acl_ids.z_aclp->z_acl_bytes);
 		}
+		dmu_tx_hold_zap(tx, zfsvfs->z_unlinkedobj, FALSE, NULL);
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: This shouldn't be needed anymore after switching to `zfs_znode_delete`.  Here and elsewhere in this patch.

File: module/zfs/zfs_vnops.c [hunk 6]
```diff
@@ -2039,13 +2054,22 @@ zfs_mkdir(struct inode *dip, char *dirname, vattr_t *vap, struct inode **ipp,
...
-	(void) zfs_link_create(dl, zp, tx, ZNEW);
+	error = zfs_link_create(dl, zp, tx, ZNEW);
+	if (error != 0) {
+		zfs_znode_delete(zp, tx);
+		zfs_dirent_unlock(dl);
+		zfs_acl_ids_free(&acl_ids);
+		dmu_tx_commit(tx);
+		iput(ZTOI(zp));
+		ZFS_EXIT(zfsvfs);
+		return (error);
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: A similar thing could be done here with an `out` label, it might make the code easier to read.

File: module/zfs/zfs_dir.c [hunk 7]
```diff
@@ -776,6 +780,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
...
 	}
+
+	value = zfs_dirent(zp, zp->z_mode);
+	error = zap_add(ZTOZSB(zp)->z_os, dzp->z_id, dl->dl_name, 8, 1,
+	    &value, tx);
+
+	/*
+	 * zap_add could fail to add the entry if it exceeds the capacity of the
+	 * leaf-block and zap_leaf_split() failed to help. In such a case bail
+	 * out.
```
[the remark this answers is not in the document]
Comment (sanjeevbagewadi) [reply module/zfs/zfs_dir.c]: Done

File: module/zfs/zfs_dir.c [hunk 8]
```diff
@@ -776,6 +780,28 @@ zfs_link_create(zfs_dirlock_t *dl, znode_t *zp, dmu_tx_t *tx, int flag)
...
+	 */
+	if (error != 0) {
+		if (!(flag & ZRENAMING) && !(flag &ZNEW)) {
```
[the remark this answers is not in the document]
Comment (sanjeevbagewadi) [reply module/zfs/zfs_dir.c]: Done

File: module/zfs/zfs_vnops.c [hunk 9]
```diff
@@ -1445,10 +1446,23 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
...
+			 * Delete the node created above.
+			 */
+			zfs_znode_delete(zp, tx);
+			zfs_dirent_unlock(dl);
+			zfs_acl_ids_free(&acl_ids);
+			dmu_tx_commit(tx);
+			iput(ZTOI(zp));
+			ZFS_EXIT(zfsvfs);
+			return (error);
+		}
```
[the remark this answers is not in the document]
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Let me give that shot.

File: module/zfs/zfs_vnops.c (hunk 5 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Makes sense. Will take care of it.

File: module/zfs/zfs_vnops.c [hunk 10]
```diff
@@ -3685,6 +3709,9 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 				VERIFY3U(zfs_link_destroy(tdl, szp, tx,
 				    ZRENAMING, NULL), ==, 0);
 			}
+		} else {
+			VERIFY3U(zfs_link_create(tdl, tzp, tx, ZRENAMING),
+			    ==, 0);
```
[the remark this answers is not in the document]
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Oops.. will fix it.

**Comment by sanjeevbagewadi** [comment]
Thank you Brian for the review comments. I am working on them and the test case.

**Comment by behlendorf** [comment]
@sanjeevbagewadi in or to make reviewing this easier could you please squash your commits together and rebase the entire PR against master.  You'll need to force update the branch on Github which is fine.

File: module/zfs/zfs_vnops.c [hunk 11]
```diff
@@ -3856,25 +3878,34 @@ zfs_symlink(struct inode *dip, char *name, vattr_t *vap, char *link,
...
 	 */
-	(void) zfs_link_create(dl, zp, tx, ZNEW);
-
-	if (flags & FIGNORECASE)
-		txtype |= TX_CI;
-	zfs_log_symlink(zilog, tx, txtype, dzp, zp, name, link);
+	error = zfs_link_create(dl, zp, tx, ZNEW);
+	if (error != 0) {
+		zfs_znode_delete(zp, tx);
+		goto out;
```
Comment (ahrens) [inline module/zfs/zfs_vnops.c]: you don't need the `goto` here

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 12]
```diff
@@ -0,0 +1,56 @@
+#!/bin/ksh -p
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: You're going to need to add to add this test case to the `casenorm` section of the `tests/runfiles/linux.run`.  This is why it wasn't run by the automated testing.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+#
+
+. $STF_SUITE/tests/functional/casenorm/casenorm.kshlib
+
+# DESCRIPTION:
+# For the filesystem with casesensitivity=mixed, normalization=none,
+# when multiple files which the same name (differing only in case) created, the
+# number of files is limited to what can fit in a fatzap leaf-block. And beyond
+# that, it fails with ENOSPC. Ensure that the create fails gracefully and does
+#  not hit and assert and panic.
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: [nit] extra leading whitespace
s/not hit and assert and panic./not hit an assert and panic./

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+#
+#
+# Copyright 2018 Nutanix, Inc.  All rights reserved.
+#
+
+. $STF_SUITE/tests/functional/casenorm/casenorm.kshlib
+
+# DESCRIPTION:
+# For the filesystem with casesensitivity=mixed, normalization=none,
+# when multiple files which the same name (differing only in case) created, the
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: s/which/with/
s/created/are created/

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+log_onexit cleanup
+log_assert "With mixed mode: ensure create fails with ENOSPC beyond a certain limit"
+
+create_testfs "-o casesensitivity=mixed -o normalization=none"
+
+names='{a,A}{b,B}{c,C}{d,D}{e,E}{f,F}{g,G}{h,H}{i,I}{j,J}{k,K}{l,L}'
+for name in $names; do
+        create_file $name
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Might be Github but it looks like there's an indent issue here.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+	ret=$?
+	log_note "Created file $name status $ret"
+	if (($ret != 0)); then
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: It would be nice if we could do this in a way where the test checks for ENOSPC and just any error.  That way we know we're failing for the right reason, that said I don't think it's critical.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+		# It is sufficient to test that create fails and does not cause a panic.
+		break;
+	fi
+done
+
+destroy_testfs
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: This could be removed since the onexit handler always runs.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+log_pass "CM-not-UN FS: lookup succeeds only if using exact name form"
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Please update the message here to reflect what's being tested.

File: module/zfs/zap.c [hunk 19]
```diff
@@ -850,10 +854,29 @@ fzap_add_cd(zap_name_t *zn,
 	if (err == 0) {
 		zap_increment_num_entries(zap, 1, tx);
 	} else if (err == EAGAIN) {
+		/*
+		 * If the last two expansions did not help, there is no point
+		 * trying to expand again
+		 */
+		if (expand_retries > MAX_EXPAND_RETRIES && prev_l == l) {
+			err = ENOSPC;
```
Comment (behlendorf) [inline module/zfs/zap.c]: You should use `err = SET_ERROR(ENOSPC);` here, similar to what's done a few lines up for `EEXIST`.

File: module/zfs/zap_micro.c [hunk 20]
```diff
@@ -363,6 +363,43 @@ mze_find_unused_cd(zap_t *zap, uint64_t hash)
...
+ * 3 chunks for names + 1 chunk for value.
+ */
+#define	MZAP_ENT_CHUNKS	(1 + ZAP_LEAF_ARRAY_NCHUNKS(MZAP_NAME_LEN) + \
+	ZAP_LEAF_ARRAY_NCHUNKS(sizeof (uint64_t)))
+
+/*
+ * Check if the current entry keeps the colliding entries under the fatzap leaf
+ * size.
+ *
+ * This is important to ensure that the microzap is upgradable to fatzap.
```
Comment (behlendorf) [inline module/zfs/zap_micro.c]: [nit] I'm not sure this needs to be its own paragraph, you could include it after the sentence above.

File: module/zfs/zfs_vnops.c [hunk 21]
```diff
@@ -1446,10 +1447,20 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
 		}
 		zfs_mknode(dzp, vap, tx, cr, 0, &zp, &acl_ids);
 
+		error = zfs_link_create(dl, zp, tx, ZNEW);
+		if (error != 0) {
+			/*
+			 * Delete the node created above.
+			 */
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: [nit]  I'd drop the comment here or update it to be clear why `zp` from `zfs_mknode` needs to be destroyed.

File: module/zfs/zfs_vnops.c [hunk 22]
```diff
@@ -3686,6 +3706,8 @@ zfs_rename(struct inode *sdip, char *snm, struct inode *tdip, char *tnm,
 				VERIFY3U(zfs_link_destroy(tdl, szp, tx,
 				    ZRENAMING, NULL), ==, 0);
 			}
+		} else {
+			VERIFY0(zfs_link_create(tdl, tzp, tx, 0));
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: Why is this only needed now?  This looks like it may address a potential existing issue.

File: module/zfs/zfs_vnops.c [hunk 23]
```diff
@@ -3856,25 +3878,34 @@ zfs_symlink(struct inode *dip, char *name, vattr_t *vap, char *link,
...
+			txtype |= TX_CI;
+		zfs_log_symlink(zilog, tx, txtype, dzp, zp, name, link);
 
-	zfs_inode_update(dzp);
-	zfs_inode_update(zp);
+		zfs_inode_update(dzp);
+		zfs_inode_update(zp);
+	}
 
+out:
```
Comment (behlendorf) [inline module/zfs/zfs_vnops.c]: Or the `out:`.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+# For the filesystem with casesensitivity=mixed, normalization=none,
+# when multiple files which the same name (differing only in case) created, the
+# number of files is limited to what can fit in a fatzap leaf-block. And beyond
+# that, it fails with ENOSPC. Ensure that the create fails gracefully and does
+#  not hit and assert and panic.
+#
+# STRATEGY:
+# 1. Create files with same name but varying in case.
+#    E.g. 'abcdefghijklmnop', 'Abcdefghijklmnop', 'ABcdefghijklmnop' etc.
+#    The create would fail with ENOSPC.
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Since this change touches several system calls we should extent the test case to verify they behave as expected too.  Specifically let's check rename, symlinks, and mkdir.

File: module/zfs/zfs_vnops.c (hunk 11 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Oops... makes sense.

File: module/zfs/zap.c (hunk 19 above)
Comment (sanjeevbagewadi) [reply module/zfs/zap.c]: Done

File: module/zfs/zap_micro.c (hunk 20 above)
Comment (sanjeevbagewadi) [reply module/zfs/zap_micro.c]: Deleting that line as it is no longer correct. As per Matt's suggestion, we now trigger an upgrade to fatzap when the number of chunks exceed the limit.

Without this change we were failing the zap_add() for a smaller number if it was a microzap.
With upgrade, we would first upgrade to fatzap and that would allow one to add more entries with colliding hash. And eventually zap_add() to fatzap would fail when the colliding entry can longer fit in the leaf-block.

File: module/zfs/zfs_vnops.c (hunk 21 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Have reworded the comment. If does not serve the purpose, I am fine with deleting it.

File: module/zfs/zfs_vnops.c (hunk 22 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: You are right, this seems like an existing issue. But, it is mostly a remote/unlikely case for the following reasons :
- We are deleting an entry on line:3677 and adding the same entry back on line:3680.
   Hence, i don't think it would fail.  Also, since this thread holds, the zfs_dirlock(), no other
   thread can race us.

However, as a safety measure, decided to add this code here.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 12 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Ah ! I missed that.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 14 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Done.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 13 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: Done.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 15 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: There was an extra space. Deleted it.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 18 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: oops.. slip up. Fixed it.

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh (hunk 16 above)
Comment (sanjeevbagewadi) [reply tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: will try to address this. The exit code does not indicate ENOSPC. Will need to rely on the error message.

File: module/zfs/zfs_vnops.c (hunk 22 above)
Comment (tuxoko) [reply module/zfs/zfs_vnops.c]: @sanjeevbagewadi 
I'll copy me gerrit comment here:
> You haven't add 'if (tzp)'
>
>Also, the zfs_link_destroy above will set z_unlinked to true, which will cause this to fail. And it will call zfs_unlinked_add. You'll need to revert those changes.

I don't feel we need to solve this in this patch, since this path is probably not very likely, and it's an existing issue. Maybe we should just have a warning here, and solve it later.

File: module/zfs/zfs_vnops.c [hunk 25]
```diff
@@ -1446,10 +1447,21 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
 		}
 		zfs_mknode(dzp, vap, tx, cr, 0, &zp, &acl_ids);
 
+		error = zfs_link_create(dl, zp, tx, ZNEW);
+		if (error != 0) {
+			/*
+			 * Since, we failed to add the directory entry for it,
+			 * delete the newly created dnode.
+			 */
+			zfs_znode_delete(zp, tx);
```
Comment (tuxoko) [inline module/zfs/zfs_vnops.c]: @sanjeevbagewadi
I'll copy me gerrit comment here:
>iput_final won't evict the inode unless you drop the link count to zero or call remove_inode_hash to unhash it. But if you just drop the link count, you will still have a window that might have obj id reuse issue I mentioned above.
>
>So the best way I think is to call remove_inode_hash.

File: module/zfs/zfs_vnops.c (hunk 22 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: Agreed. I had overlooked the fact the zfs_link_create() will not solve the problem. For now I will add an ASSERT() here. And we can take that up separately.

File: module/zfs/zfs_vnops.c (hunk 25 above)
Comment (sanjeevbagewadi) [reply module/zfs/zfs_vnops.c]: makes sense. Will add remove_inode_hash().

**Comment by sanjeevbagewadi** [comment]
@behlendorf , I have updated the test case to cover the different object types (files, dir, symlink, hardlinks) and for each object type exercise the create and rename operation. Kindly take a look.

@ahrens , for the corner case in zfs_rename() for now, I have added an assert() there. If you agree we could take that up as a separate case.

@tuxoko , I have now added remove_inode_hash().

File: tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh [hunk 26]
```diff
@@ -0,0 +1,137 @@
...
+# STRATEGY:
+# Repeat the below steps for objects: files, directories, symlinks and hardlinks
+# 1. Create objects with same name but varying in case.
+#    E.g. 'abcdefghijklmnop', 'Abcdefghijklmnop', 'ABcdefghijklmnop' etc.
+#    The create should fail with ENOSPC.
+# 2. Create an object with name 'tmp_obj' and try to rename it to name that we
+#    failed to add in step 1 above.
+#    This should fail as well.
+
+set -x
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/casenorm/mixed_create_failure.ksh]: nit: Let's drop the `set -x` for the final version.
